source/validate_decorations.cpp

```diff
@@ -48,7 +65,49 @@ spv_result_t ValidateDecorations(ValidationState_t& vstate) {
     }
   }
 
-  // TODO: Add more decoration validation code here.
+  for (uint32_t entry_point : vstate.entry_points()) {
+    const auto& interfaces = vstate.entry_point_interfaces(entry_point);
+    int num_builtin_inputs = 0;
+    int num_builtin_outputs = 0;
+    for (auto interface : interfaces) {
+      Instruction* var_instr = vstate.FindDef(interface);
+      if (SpvOpVariable != var_instr->opcode()) {
+        return vstate.diag(SPV_ERROR_INVALID_ID)
+               << "Interfaces passed to OpEntryPoint must be of type "
+                  "OpTypeVariable. Found Op"
+               << spvOpcodeString(static_cast<SpvOp>(var_instr->opcode()))
+               << ".";
+      }
+      const uint32_t ptr_id = var_instr->word(1);
+      Instruction* ptr_instr = vstate.FindDef(ptr_id);
+      // It is guaranteed (by validator ID checks) that ptr_instr is
+      // OpTypePointer. Word 3 of this instruction is the type being pointed to.
+      const uint32_t type_id = ptr_instr->word(3);
+      Instruction* type_instr = vstate.FindDef(type_id);
+      if (type_instr && SpvOpTypeStruct == type_instr->opcode() &&
+          isBuiltInStruct(type_id, vstate)) {
+        const auto storage_class =
+            static_cast<SpvStorageClass>(var_instr->word(3));
+        if (storage_class != SpvStorageClassInput &&
```

Comment:
This error is independent of whether the current thing has builtins.  So pull this check outside of the "if" at line 87.